rfcs/45-stageless.md

```diff
@@ -0,0 +1,761 @@
+    - Add `state_equals` function that generates an `IntoRunCondition`-compatible closure for checking "if active".
+- Misc.
+  - Rename "system chaining" to "system piping" so we can use "chain" to describe the macro/type that applies sequential ordering constraints.
+- Port the rest of the engine and examples to the new API.
+
+Given the massive scope, that sounds relatively straightforward!
+However, doing so will break approximately the entire engine, and tests will not pass again until everything has been ported.
+
+Let's explore some implementation details.
+
+### Scheduling
+
+Systems and their scheduling information are bundled and stored.
+The actual dependency graph construction is deferred (which means scheduling methods are order-independent).
+
+```rust
+enum Order {
+    Before,
+    After
+}
+
+struct Config {
+    sets: HashSet<BoxedSystemLabel>,
+    edges: Vec<(Order, BoxedSystemLabel)>,
+    conditions: Vec<BoxedRunCondition>,
+}
+
+struct ConfiguredSystem {
+    system: BoxedSystem,
+    config: Config,
+    name: Option<BoxedSystemLabel>,
+}
+
+struct ConfiguredSet {
+    set: BoxedSystemLabel,
+    config: Config,
+}
+
+trait IntoConfiguredSystem<Params> {
+    fn configure(self) -> ConfiguredSystem;
+    fn configure_with(self, config: Config) -> ConfiguredSystem;
+    /// Configures the system to run before the system or set given by `label`.
+    fn before<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystem;
+    /// Configures the system to run after the system or set given by `label`.
+    fn after<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystem;
+    /// Configures the system to run as part of `set`.
+    fn in_set(self, set: impl SystemLabel) -> ConfiguredSystem;
+    /// Configures the system to run only if `condition` returns `true`.
+    fn run_if<P>(self, condition: impl IntoRunCondition<P>) -> ConfiguredSystem;
+    /// Sets `name` as the unique name for this system.
+    fn named(self, name: impl SystemLabel) -> ConfiguredSystem;
+}
+
+trait IntoConfiguredSet {
+    fn configure(self) -> ConfiguredSet;
+    fn configure_with(self, config: Config) -> ConfiguredSet;
+    /// Configures the system set to run before the system or set given by `label`.
+    fn before<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSet;
+    /// Configures the system set to run after the system or set given by `label`.
+    fn after<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSet;
+    /// Configures the system set to run as part of `set`.
+    fn in_set(self, set: impl SystemLabel) -> ConfiguredSet;
+    /// Configures the system set to run only if `condition` returns `true`.
+    fn run_if<P>(self, condition: impl IntoRunCondition<P>) -> ConfiguredSet;
+}
+```
+
+### Scheduling errors
+
+There are several reasons an app may error due to improper scheduling. Most are related to graph solvability.
+
+1. Your dependencies contain a loop or cycle.
+2. Your hierarchy contains a loop or cycle.

```

Comment:
I think this is referring to the idea of the sets themselves containing a loop or cycle. I agree that this should be merged down though.

Reply:
Actually, the term “hierarchies” is mentioned before, in the “Scheduling and terminology overview”, after the glossary.

It should be made clear though in the user-facing explanation that by default, it is required that dependencies between sets or systems should happen at the same hierarchy level.

In other words, the dependency graph can be described as a nested graph (or multi-level graph), where a set is a node containing a graph by its own.

So:

- (1) and (2) looks the same
- (3) looks like (5)
- The sentence inside the parentheses on (3) makes no sense
- (4) is not clear at all

Reply:
> In other words, the dependency graph can be described as a nested graph (or multi-level graph), where a set is a node containing a graph by its own.

Yeah, nesting sets creates multi-level dependency graphs and I guess I've just been calling the vertical slices "hierarchies." Their graphs and not trees since you can add a system/set to more than one set.

There's no relevant structs and algorithms for multi-level graphs, so in my draft impl, the hierarchical relations (`.in_set`) and dependency relations (`.before`, `.after`) between nodes are stored in separate graphs.

And yep, the pairs of (1) + (2) and (3) + (5) are the same, one for each kind of graph.